app/scripts/modules/core/securityGroup/securityGroupReader.service.ts

```diff
@@ -0,0 +1,412 @@
+
+      return this.attachSecurityGroups(application, nameBasedSecurityGroups, false);
+    });
+  }
+
+  private addStackToSecurityGroup(securityGroup: ISecurityGroup): void {
+    const nameParts: IComponentName = this.namingService.parseSecurityGroupName(securityGroup.name);
+    securityGroup.stack = nameParts.stack;
+  }
+
+  private attachSecurityGroups(application: Application,
+                               nameBasedSecurityGroups: ISecurityGroup[],
+                               retryIfNotFound: boolean): ng.IPromise<any[]> {
+
+    let data: ISecurityGroup[] = [];
+    let notFoundCaught = false;
+    if (nameBasedSecurityGroups) {
+      // reset everything
+      application.getDataSource('securityGroups').data = [];
+      const nameBasedGroups: ISecurityGroupProcessorResult =
+        this.addNameBasedSecurityGroups(application, nameBasedSecurityGroups);
+      notFoundCaught = nameBasedGroups.notFoundCaught;
+      if (!nameBasedGroups.notFoundCaught) {
+        data = nameBasedGroups.securityGroups;
+      }
+    } else {
+      // filter down to empty (name-based only) security groups - we will repopulate usages
+      data = application
+        .getDataSource('securityGroups')
+        .data
+        .filter((group: ISecurityGroup) => !group.usages.serverGroups.length && !group.usages.loadBalancers.length);
+    }
+
+    if (!notFoundCaught) {
+
+      const loadBalancerSecurityGroups: ISecurityGroupProcessorResult =
+        this.addLoadBalancersSecurityGroups(application);
+      notFoundCaught = loadBalancerSecurityGroups.notFoundCaught;
+      if (!notFoundCaught) {
+
+        data = data.concat(loadBalancerSecurityGroups.securityGroups.filter((sg: any) => !data.includes(sg)));
+        const serverGroupSecurityGroups: ISecurityGroupProcessorResult = this.addServerGroupSecurityGroups(application);
+        notFoundCaught = serverGroupSecurityGroups.notFoundCaught;
+        if (!notFoundCaught) {
+          data = data.concat(serverGroupSecurityGroups.securityGroups.filter((sg: any) => !data.includes(sg)));
+        }
+      }
+    }
+
+    data = uniq(data);
+    if (notFoundCaught && retryIfNotFound) {
+      this.$log.warn('Clearing security group cache and trying again...');
+      return this.clearCacheAndRetryAttachingSecurityGroups(application, nameBasedSecurityGroups);
+
+    } else {
+      data.forEach((sg: ISecurityGroup) => this.addStackToSecurityGroup(sg));
+      return this.$q.all(data.map((sg: ISecurityGroup) => this.transformer.normalizeSecurityGroup(sg)))
+        .then(() => this.$q.when(data));
+    }
+  }
+
+  constructor(private $log: ng.ILogService,
+              private $q: ng.IQService,
+              private searchService: any,
+              private namingService: NamingService,
+              private API: Api,
+              private cache: InfrastructureCacheService,
+              private transformer: SecurityGroupTransformerService,
+              private delegate: any) {
+  }
```

Comment:
I was under the impression we wanted to explicitly declare constructors `public` - is that just for components?